// @big-sister/core/startCollector.ts
import { Extension, Identifier, Presence, Spy, Store } from '@big-sister'
import log from './utility/logging'

export default function startCollector(
  spy: Spy,
  store: Store,
  extensions: Extension[],
): Collector {
  const cache = new Map<Identifier, boolean>()
  const presence$ = extensions.reduce((p$, e) => e.transformPresence(p$), spy.presence$)
  presence$.subscribe(async presence => {
    const { presentIdentifiers, timestamp } = presence
    const absentIdentifiers = new Set(cache.keys())
    await Promise.all(presentIdentifiers.map(async identifier => {
      if (cache.has(identifier)) {
        log.verbose(`'${identifier}' stays present`)
        await store.stayPresent(identifier, timestamp)
        absentIdentifiers.delete(identifier)
      } else {
        log.verbose(`'${identifier}' becomes present`)
        await store.becomePresent(identifier, timestamp)
        cache.set(identifier, true)
      }
    }))
    await Promise.all([...absentIdentifiers].map(async identifier => {
      log.verbose(`'${identifier}' becomes absent`)
      await store.becomeAbsent(identifier, timestamp)
      cache.set(identifier, false)
    }))
  })
  log.info('Collector started')
  return { getCurrentPresence }

  function getCurrentPresence(): Presence {
    const presence: Presence = { presentIdentifiers: [], timestamp: new Date() }
    for (const [identifier, isPresent] of cache) {
      if (isPresent) presence.presentIdentifiers.push(identifier)
    }
    return presence
  }
}

/**
 * Collects presence data.
 */
export interface Collector {

  /**
   * Determine the current presence.
   */
  getCurrentPresence(): Presence
}
